designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,784 @@
+
+#### Extending From `eslint:recommended` and `eslint:all`
+
+Both `eslint:recommended` and `eslint:all` can be represented as strings in a config array. For example:
+
+```js
+exports.config = [
+    "eslint:recommended",
+    require("eslint-config-standard"),
+    require("@me/eslint-config"),
+    {
+        files: ["*.js"],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+This config first extends `eslint:recommended` and then continues on to extend other configs.
+
+#### Setting the Name of Shareable Configs
+
+For shareable configs, specifying a `name` property for each config they export helps ESLint to output more useful error messages if there is a problem. The `name` property is a string that will be used to identify configs to help users resolve problems. For example, if you are creating `eslint-config-example`, then you can specify a `name` property to reflect that:
+
+```js
+module.exports = {
+    name: "eslint-config-example",
+
+    // other info here
+};
+```
+
+It's recommended that the shareable config provide a unique name for each config that is exported.
+
+#### Disambiguating Shareable Configs With Common Dependencies
+
+Today, shareable configs that depend on plugin rules must specify the plugin as a peer dependency and then either provide a script to install those dependencies or ask the user to install them manually.
+
+With this design, shareable configs can specify plugins as direct dependencies that will be automatically installed with the shareable config, improving the user experience of complex shareable configs. This means its possible for multiple shareable configs to depend on the same plugin and, in theory, depend on different versions of the same plugin. In general, npm will handle this directly, installing the correct plugin version at the correct level for each shareable config to `require()`. For example, suppose there are two shareable configs, `eslint-config-a` and `eslint-config-b` that both rely on the `eslint-plugin-example` plugin, but the former relies on 1.0.0 and the latter relies on 2.0.0. npm will install those plugins like this:
+
+```
+your-project
+├── eslint.config.js
+└── node_modules
+  ├── eslint
+  ├── eslint-config-a
+  | └── node_modules
+  |   └── eslint-plugin-example@1.0.0
+  └── eslint-config-b
+    └── node_modules
+      └── eslint-plugin-example@2.0.0
+```
+
+The problem comes when the shareable configs try to use the default namespace of `eslint-plugin-example` for its rules, such as:
+
+```js
+module.exports = {
+    ruledefs: {
+        ...require("eslint-plugin-example").ruledefs
+    }
+};
+```
+
+If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.
```

Comment:
To be clear, is an error thrown when two shareable configs with conflicting ruledefs are *merged*, or simply whenever they both appear in a final config at all (even if they apply to disjoint sets of files)?

Based on https://github.com/eslint/rfcs/pull/10#issuecomment-460706972 it seems like it's the latter, but it might be worthwhile to clarify this.

Reply:
It will happen when they are merged, I'll update that.

Reply:
In https://github.com/eslint/rfcs/pull/10#issuecomment-460706972 you mentioned that rule IDs are unique within a particular config. If conflict detection only happens when configs are merged, it seems like rule IDs wouldn't be unique for this config:

```js
exports.config = [
  {
    files: ["*.js"],
    ruledefs: {
      foo: RULESET_ONE
    }
  },
  {
    files: ["*.ts"],
    ruledefs: {
      foo: RULESET_TWO
    }
  }
]
```

There are no files that match both configs, so the two configs would never be merged and no conflict would be detected. However, there would be a duplicate "foo" ruledef, and possibly duplicate rule IDs, which would cause problems for https://github.com/eslint/rfcs/pull/10. I think it would be better for an error to be thrown for this case.

Reply:
That's a good point, we can certainly detect that during the normalization phase.